api/tests/unit/segments/test_unit_segments_models.py

```diff
@@ -167,3 +169,247 @@ def test_condition_get_update_log_message(segment, segment_rule, mocker):
+    assert new_child_rule2.type == child_rule2.type
+
+    new_child_rule3, new_child_rule4 = list(new_parent_rule2.rules.all())
+
+    assert new_child_rule3.type == child_rule3.type
+    assert new_child_rule4.type == child_rule4.type
+
+    assert (
+        len(new_parent_rule1.conditions.all())
+        == len(parent_rule1.conditions.all())
+        == 2
+    )
+
+    new_condition1, new_condition2 = list(new_parent_rule1.conditions.all())
+
+    assert new_condition1.property == condition1.property
+    assert new_condition1.operator == condition1.operator
+    assert new_condition1.value == condition1.value
+    assert new_condition1.created_with_segment is condition1.created_with_segment
+
+    assert new_condition2.property == condition2.property
+    assert new_condition2.operator == condition2.operator
+    assert new_condition2.value == condition2.value
+    assert new_condition2.created_with_segment is condition2.created_with_segment
+
+    assert (
+        len(new_child_rule1.conditions.all()) == len(child_rule1.conditions.all()) == 1
+    )
+    new_condition3 = new_child_rule1.conditions.first()
+
+    assert new_condition3.property == condition3.property
+    assert new_condition3.operator == condition3.operator
+    assert new_condition3.value == condition3.value
+    assert new_condition3.created_with_segment is condition3.created_with_segment
+
+    # Since the conditions have been asserted at multiple levels,
+    # we leave the rest to basic size assertions.
+    assert (
+        len(new_child_rule2.conditions.all()) == len(child_rule2.conditions.all()) == 2
+    )
+    assert (
+        len(new_child_rule3.conditions.all()) == len(child_rule3.conditions.all()) == 1
+    )
+    assert (
+        len(new_child_rule4.conditions.all()) == len(child_rule4.conditions.all()) == 1
+    )
+
+
+def test_manager_returns_only_highest_version_of_segments(
+    segment: Segment,
+) -> None:
+    # Given
+    # The built-in segment fixture is pre-versioned already.
+    assert segment.version == 2
+    assert segment.version_of == segment
+
+    new_segment = segment.deep_clone()
+    assert new_segment.version == 2
+    assert segment.version == 3
+
+    # When
+    queryset1 = Segment.objects.filter(id=new_segment.id)
+    queryset2 = Segment.all_objects.filter(id=new_segment.id)
+    queryset3 = Segment.objects.filter(id=segment.id)
+    queryset4 = Segment.all_objects.filter(id=segment.id)
+
+    # Then
+    assert not queryset1.exists()
+    assert queryset2.first() == new_segment
+    assert queryset3.first() == segment
+    assert queryset4.first() == segment
+
+
+def test_deep_clone_of_segment_with_sub_rule(
+    project: Project,
+    feature: Feature,
+) -> None:
+    # Given
+    segment = Segment.objects.create(
+        name="SpecialSegment",
+        description="A lovely, special segment.",
+        project=project,
+        feature=feature,
+    )
+
+    sub_rule = SegmentRule.objects.create(type=SegmentRule.ALL_RULE)
```

Comment:
I'm a little unsure what's going on in this test. This `sub_rule` is not valid since it doesn't have a parent segment or rule and `sub_rule` to me indicates that it's a rule underneath something, whereas below we're using it as a parent of what we define (via the comment) as a parent rule.

Reply:
Yeah I had it backwards in my head when I was writing it. I've fixed this locally. Either way the test is of a misbehaving rule, but I've added the segment to the rule before also adding the rule which has two parents (needed for the test).